# ChatGPT Mental Support simulation

In this notebook we will simulate how chatGPT will respond to request from people with mental issues, problems and worries, that will be expressed plain in conversation.
This responses will be compared to responses from mental counselor to see how is chatGPT doing this.

### Dataset

This simulation is based on this dataset:

[Mental health Counseling Conversations dataset](https://www.kaggle.com/datasets/thedevastator/nlp-mental-health-conversations)

### Dataset structure

| Column   | Description | 
|----------|----------|
| Context  | Request, phrase, query sentence from the patient, that seems to have mental troubles, problem, disorders of different kind        |  
| Response | Response, that mental counselor is supposed to give. For one ```Context``` there is from 1 to 47 different responses (mean 2.5) | 


### Comparison

Comparison between ChatGPT response will be done with special prompt, resulting in score [0-100], and be done comparing one response from ChatGRT to all (one by one) responses from the dataset. New dataset will chatGPT responses and scores will be build, and then explored



# Notebook for simulation

Exploration of data and cleaning,
Probes with ChatGPT are places in the [preparation notebook](notebook_preparation_work.ipynb)

Here only functions to make data ready, and helper functions are presented

## Preparing the data

The dataset on the link above is available to download, so here it is already present in the ```dataset/NLP_Mental_Health_Conversations.csv```

In [2]:
#!pip install panda
# !pip install --upgrade tiktoken

In [3]:
import pandas as pd
import numpy as np

C:\Users\sirius\AppData\Local\Temp\ipykernel_26728\2162656668.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [4]:
CONVERSATIONS_DATASET_FILENAME = 'dataset/NLP_Mental_Health_Conversations.csv'

conversation_dataset = pd.read_csv(CONVERSATIONS_DATASET_FILENAME)

In [5]:
conversation_dataset.columns = conversation_dataset.columns.str.lower().str.replace(' ', '_')

### Removing duplicates and zero entries

In [6]:
conversation_dataset.dropna(subset=['response'], inplace=True)
conversation_dataset.drop_duplicates(subset=['response'], inplace=True)
conversation_dataset.reset_index(drop=True, inplace=True)

So there is duplicate entries, but they differ in heading or trailing spaces.
Let's fix this, and remove duplicates once more.

In [7]:
conversation_dataset['response'] = conversation_dataset['response'].apply(
    lambda x: x.strip())

In [8]:
conversation_dataset.drop_duplicates(subset=['response'], inplace=True)
conversation_dataset.reset_index(drop=True, inplace=True)

### Errors in dataset

Working with the dataset I found some irrelevant entries. So we should just delete them


In [9]:
error_context = [" such as not enough sleep",
                 "Can a counselor take sides with one parent and allow a parent to order the child to tell the counselor"]

conversation_dataset = conversation_dataset[~conversation_dataset['context'].str.startswith(
    tuple(error_context))]

In [10]:
conversation_dataset.describe()

context  \
count                                                2049   
unique                                                841   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2049  
unique                                               2049  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

In [11]:
context_unique = conversation_dataset.groupby(
    'context').size().reset_index(name='count')

In [12]:
context_unique.sort_values('count', ascending=False).head(30)

context  count
286  I have so many issues to address. I have a his...     47
78   How do you know you have the right therapist f...     41
545  Is it normal for people to cry during therapy,...     38
523  I've gone to a couple therapy sessions so far ...     26
828  When my daughter is stressed about a silly thi...     23
461  I'm going through some things with my feelings...     23
80     How does a person start the counseling process?     22
79   How does a counselor decide when to end counse...     20
31   Does counseling really do anything that can he...     20
73   How do I decide if I should end my relationshi...     15
41   Every winter I find myself getting sad because...     15
695  My mother takes care of niece whom my sister a...     13
353  I start counseling/therapy in a few days (I'm ...     12
306  I just took a job that requires me to travel f...     11
396  I was raped a couple months ago, Since then, a...     11
823  When I'm in large crowds I get angry and I jus...     11
643  My fiancé and I have been together for 3 years...     11
311  I keep having these random thoughts that I don...     11
213  I have a friend that who I used to be in a rel...     10
693  My mother has Alzheimer's and she has become s...     10
484  I'm obsessing about a terrible breakup. Everyt...      9
624  My daughter is in later elementary school. She...      9
303  I just lost my grandpa and i'm having a rough ...      9
465  I'm having relationship problems and I want to...      9
813                What makes a healthy marriage last?      9
154  I don't know what to say. I have never really ...      9
566  I’ve been in a long distance relationship with...      9
246  I have been married for 11 years. Within the p...      9
16   After he got home from the hospital he was ang...      9
561  I’m facing severe depression and anxiety and I...      9

In [13]:
context_unique['count'].describe()

count    841.000000
mean       2.436385
std        3.538705
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max       47.000000
Name: count, dtype: float64

In [14]:
context_unique['count'].value_counts()

count
1     452
2     182
3      78
4      40
5      23
6      18
9      12
7      10
8       6
11      5
23      2
20      2
15      2
10      2
38      1
26      1
41      1
12      1
47      1
22      1
13      1
Name: count, dtype: int64

In [15]:
CONTEXT_UNIQUE_FILENAME = 'working_data/context_unique.csv'

context_unique.to_csv(CONTEXT_UNIQUE_FILENAME)

### Data is cleaned and ready for simulation

So we have 845 unique requests, with a mean of 2.45 responses for each, and 452 having one response, and max of 47 responses, but only 10 requests have more then 10 responses.

### Conclusions

**Maybe for simulation we should use only requests with 2 or more responses, so we have some variety in approved responses to compare with. If we take request with 2 or more we will have 393 requests, if we take with 3 or more - 209 requests. 209 may be enough for our simulation. And having only one response will make it possible for incorrect scoring of ChatGPT output. Because mental issues is delicate matter, so several different responses may serve well for specific case.**


## ChatGPT

We follow the standard way of using OpenAI API for reaching ChatGPT
The below code uses ```.env``` file in this directory, which is not included to the repository.
```.env```  which must contain:
```
OPENAI_API_KEY=...
```
This API key to be used to interact with OpenAI API

If you want to use this code, you must [get your own API key from OpenAI](https://platform.openai.com/api-keys)

Moreover you will need to make a payment to your ChatGPT account, to be able to use it.

This code is based on the [course by DeepLearning.AI](https://learn.deeplearning.ai/courses/langchain/lesson/2/models%2C-prompts-and-parsers)

In [16]:
# !pip install python-dotenv
# !pip install openai

In [17]:
import os
import openai

from dotenv import load_dotenv, find_dotenv
_ = load_dotenv(find_dotenv())  # read local .env file

## ChatGPT client and initial functions

In [18]:
GPT_MODEL = 'gpt-3.5-turbo-0125'
#
# Pricing
# Model	Input	Output
# gpt-3.5-turbo-0125	$0.0005 / 1K tokens	$0.0015 / 1K tokens
#
# Updated GPT 3.5 Turbo
# The latest GPT-3.5 Turbo model with higher accuracy at responding in requested formats
# and a fix for a bug which caused a text encoding issue for non-English language function calls.
# Returns a maximum of 4,096 output tokens.
#

In [19]:
client = openai.OpenAI(

    api_key=os.environ.get("OPENAI_API_KEY"),
)

In [20]:
chat_completion = client.chat.completions.create(
    messages=[
        {
            "role": "user",
            "content": "Hello, this is a test",
        }
    ],
    model=GPT_MODEL,
)

# print(chat_completion.choices[0].message.content)

In [21]:
import random

# default temperature for ChatGPT. Web ChatGPT uses 0.6 all the time. varies from 0 (same answers, no randomness) to 1 (much variety, much random)
TEMPERATURE = 0.6
SPEAK_TEMPERATURE = 0.6

# Our randomness constants:

USE_STARTUP_DIALOG = 0.25

# if we use startup dialog, we may have up to 4 preceding phrases.

# So len 1 will occur more frequently, len 4 - much less
DIALOG_LENGTH_WEIGHTS = [0.35, 0.3, 0.2, 0.15]
DIALOG_LENGTH_OPTIONS = [2, 4, 6, 8]

starting_messages = [
    [
        {'role': 'user', 'content': "Hi"},
        {'role': 'assistant', 'content': "Hello! How can I assist you today?"},
        {'role': 'user', 'content': "Yes! How are you?"},
        {'role': 'assistant', 'content': "As an AI, I don't have feelings or emotions, but I'm here and ready to assist you with any questions or tasks you have. How can I help you today?"},
        {'role': 'user', 'content': "What do you do"},
        {'role': 'assistant', 'content': "I'm here to help with a wide range of tasks! You can ask me questions, seek information on various topics, brainstorm ideas, get assistance with writing or editing, play text-based games, generate creative content, and much more. Just let me know what you need help with, and I'll do my best to assist you!"},
        {'role': 'user', 'content': 'Yes'},
        {'role': 'assistant', 'content': "Great! If you have any specific questions or need assistance with something, feel free to ask. I'm here to help!"},
    ],
    [
        {'role': 'user', 'content': "Hello!"},
        {'role': 'assistant', 'content': "Hello! How can I assist you today?"},
        {'role': 'user', 'content': "I don't know"},
        {'role': 'assistant', 'content': "That's perfectly fine! If you have any questions, need advice, want to discuss a topic, or just want to chat, feel free to let me know. I'm here to help with anything you need."},
        {'role': 'user', 'content': "Ok. I will think about it"},
        {'role': 'assistant', 'content': "Sure thing! Take your time to think, and whenever you're ready, feel free to reach out. I'll be here to assist you whenever you need."},
        {'role': 'user', 'content': "Tell me something"},
        {'role': 'assistant',
            'content': "Sure! Did you know that the world's largest desert isn't the Sahara, but actually Antarctica? Despite being covered in ice, Antarctica is considered a desert because it receives very little precipitation, less than 2 inches (50 millimeters) per year in some areas. Its size and harsh climate make it the largest desert on Earth, even larger than the Sahara Desert."},
    ],
    [
        {'role': 'user', 'content': "Nice to see you"},
        {'role': 'assistant', 'content': "Nice to see you too! How can I assist you today?"},
        {'role': 'user', 'content': "What can you say?"},
        {'role': 'assistant', 'content': "As an AI language model, I can engage in conversation, answer questions, provide information on a wide range of topics, generate text based on prompts, offer suggestions, and more. Just let me know what you need assistance with or what you'd like to discuss!"},
        {'role': 'user', 'content': "Just tell me something"},
        {'role': 'assistant', 'content': "Sure! Did you know that honey never spoils? Archaeologists have found pots of honey in ancient Egyptian tombs that are over 3,000 years old and still perfectly edible. This remarkable longevity is due to the unique chemical composition of honey, which includes low water content and natural acidity, making it inhospitable to bacteria and microorganisms that cause food to spoil."},
        {'role': 'user', 'content': "Interesting"},
        {'role': 'assistant', 'content': "I'm glad you found it interesting! If you have any other topics you'd like to learn about or discuss, feel free to let me know. I'm here to help!"},
    ],
    [
        {'role': 'user', 'content': "How are things"},
        {'role': 'assistant', 'content': "As an AI, I don't have personal experiences, but I'm here and ready to assist you with any questions or tasks you have! How can I help you today?"},
        {'role': 'user', 'content': "I am unsure"},
        {'role': 'assistant', 'content': "It's okay to feel unsure sometimes. If there's anything specific on your mind or any decision you're facing, feel free to share, and I'll do my best to assist you. Whether it's about making a choice, seeking advice, or just discussing your thoughts, I'm here to help."},
        {'role': 'user', 'content': "yes talk to me"},
        {'role': 'assistant', 'content': "Of course! Is there anything particular on your mind that you'd like to talk about? It could be anything from personal concerns, questions about a topic, or just casual conversation. Feel free to share, and I'm here to listen and offer assistance or insights as best as I can."},
        {'role': 'user', 'content': "i don't know"},
        {'role': 'assistant', 'content': "That's completely fine. Sometimes it can be helpful to explore different topics or interests to see if anything sparks your curiosity or engages your thoughts. Is there a subject you've been curious about or something you've been meaning to learn more about? Alternatively, we can engage in a casual conversation about anything you'd like. Let me know if there's a specific direction you'd like to take our conversation or if there's anything specific you'd like to discuss."},
    ],
]

### Helper function for request - response

In [22]:
def generate_messages(phrase_request: str) -> list[dict]:
    """Using random this function will generate messages to pass to ChatGPT

    Args:
        phrase_request (str): Request phrase to put to the end of messages list

    Returns:
        list[dict]: messages list in a OpenAi format
    """
    # this is a standard option
    return_messages = [{'role': 'system', 'content': "You are a helpful assistant."},
                       ]

    # use random to chose if we are adding anything
    if random.random() < USE_STARTUP_DIALOG:
        # choose number of adding, based on provided weights
        startup_len_to_add = random.choices(
            DIALOG_LENGTH_OPTIONS, weights=DIALOG_LENGTH_WEIGHTS)[0]
        message_chain = random.randint(0, len(starting_messages)-1)
        return_messages += starting_messages[message_chain][:startup_len_to_add]

    # finally our phrase
    return return_messages+[
        {'role': 'user', 'content': phrase_request},
    ]

In [23]:
def request_gpt(messages: list[dict], temperature=TEMPERATURE, model=GPT_MODEL) -> str:
    """Sends messages to ChatGPT and returns the response

    Args:
        messages (list[dict]): list of messages in OpenAI format

    Returns:
        str: response of ChatGpt
    """
    chat_completion = client.chat.completions.create(
        messages=messages,
        model=model,
        temperature=temperature
    )

    return chat_completion.choices[0].message.content

In [24]:
def speak_with_gpt(phrase_request: str, temperature=SPEAK_TEMPERATURE) -> str:
    """Use phrase_request to generate messages with generate_messages function, send's them to ChatGPT, and return response

    Args:
        phrase_request (str): Phrase to ask

    Returns:
        str: Response of ChatGPT
    """
    return request_gpt(generate_messages(phrase_request=phrase_request), temperature=temperature)

### Helper function for comparing the results

In [25]:
# lower is better in such cases, we will get more sustainable responses
COMPARING_TEMPERATURE = 0.3

#
# note: i have made experiments with it until i get it working correctly
#

COMPARING_PROMPT_TEMPLATE_2 = """
You will need to compare two phrases and scale their likeness. 

Compare PHRASE_1 and PHRASE_2 by summarized meaning, rate similarity by scale from 0 to 100 (0 - no similarity or opposite meaning, 100 - total likeness, similar meaning).

Response with number only!


```PHRASE_1```
<<<PHRASE1###>>>
```END OF PHRASE_1```


```PHRASE_2```
<<<PHRASE2###>>>
```END OF PHRASE_2```

SCALE OF LIKENESS:
"""

In [66]:
import tiktoken

def num_tokens_from_string(string: str, encoding_name:str="cl100k_base") -> int:
    encoding = tiktoken.get_encoding(encoding_name)
    num_tokens = len(encoding.encode(string))
    return num_tokens

# This is corrected one

def generate_comparing_prompt(response1: str, response2: str, comparing_template=COMPARING_PROMPT_TEMPLATE_2) -> str:
    """Insert into COMPARING_PROMPT_TEMPLATE given request and responses

    Args:
        response1 (str): response1
        response2 (str): response2

    Returns:
        str: Prompt to send to ChatGRT for comparing
    """
    result=comparing_template.replace('<<<PHRASE1###>>>', response1).replace('<<<PHRASE2###>>>', response2)
    while num_tokens_from_string(result)>14000:
        response2=response2[:-abs(num_tokens_from_string(result)-14000)]
        result=comparing_template.replace('<<<PHRASE1###>>>', response1).replace('<<<PHRASE2###>>>', response2)
    return result


In [67]:
def compare_with_gpt(response1: str, response2: str, temperature=COMPARING_TEMPERATURE, model=GPT_MODEL) -> int:
    """Makes a prompt to compare phrase1 and phrase2, send to ChatGPT and return and integer 

    Args:
        response1 (str): response1
        response2 (str): response2

    Returns:
        int: 0..100 - scale of likeness, -1 if something goes wrong
    """
    messages = [{'role': 'system', 'content': "You are a helpful assistant. "},
                {'role': 'user', 'content': generate_comparing_prompt(
                    response1, response2)},
                ]
    response = request_gpt(messages, temperature=temperature)

    try:
        scale_number = int(response.strip())
    except ValueError:
        scale_number = -1

    return scale_number

### Labeling

we will label the requests for futher analysis

In [28]:
LABELING_TEMPERATURE = 0

LABEL_PROMPT_1 = """
Analyze the text and label the topic of it. Label must be of one or several words, and name the type of mental disorder the author of text is facing. 
Examples of topics: [ Depression, Anxiety, Eating Disorder, Drug Use, Addiction, Schizophrenia, Bipolar Disorder, Self-Esteem, Personality Disorders, .... ]
You should use common label names with low variety, so I could use this label later in my project.
Provide two labels maximum, that are more relevant.
Respond with label names only, separate different labels with coma ","!
<<<TEXT>>>
<<<<<<TEXT###>>>>>>
<<<END OF TEXT>>>

<<<LABEL>>>:
"""


def generate_labeling_prompt(text_to_label: str, comparing_template=LABEL_PROMPT_1) -> str:
    """Insert into comparing_template given text to send later to ChatGPT

    Args:
        text_to_label (str): text to label

    Returns:
        str: Prompt to send to ChatGRT for labeling
    """
    return comparing_template.replace('<<<<<<TEXT###>>>>>>', text_to_label)


def get_label_from_chatGPT(text_to_label: str, temperature=LABELING_TEMPERATURE) -> str:
    """Generate label prompt, send to ChatGPT and get label for given text

    Args:
        text_to_label (str): text to label

    Returns:
        str: label
    """
    messages = [{'role': 'system', 'content': "You are a helpful assistant. You label text with labels from a limited common set, so they could later be used to aggregate texts."},
                {'role': 'user', 'content': generate_labeling_prompt(
                    text_to_label)},
                ]
    response = request_gpt(messages, temperature=temperature)

    return response

## Simulation 

We will make a new dataset and run a simulation for filling it 

In [29]:

gpt_simulation = pd.DataFrame(columns=['request', 'label', 'ds_responses_count',
                              'chatgpt_response', 'scores', 'min_score', 'max_score', 'mean_score'])

GPT_SIMULATION_DATASET_FILENAME = 'simulation_data/gpt_simulation_01_'

In [77]:
def run_one_request(request:str,label:str,gpt_dataset=gpt_simulation,max_comparing=10):
    """Send's request_str to ChatGPT and analyze the response.

    Args:
        request_str (str): Request to send to chatGPT
        label (str): Label of request
        dataset (_type_, optional): Dataset to write results. Defaults to gpt_simulation.
        max_comparing (int, optional): To how many original dataset responses we should compare this GPTresponse. Defaults to 15.
        
    """
    current_ds_len=len(gpt_dataset)
    ds_responses_count=context_unique[context_unique['context']==request]['count'].iloc[0]
    ds_responses=conversation_dataset[conversation_dataset['context']==request]['response']
    if len(ds_responses)!=ds_responses_count:
        print("ERROR WITH COUNT MISMATCH")
        return
    print('SIM ('+str(current_ds_len)+') ['+request[:15]+'] ', end='')
    chatgpt_response=speak_with_gpt(request, temperature=SPEAK_TEMPERATURE)
    print(' -> ','{'+chatgpt_response[:100]+'} :: [', end='')
    scores=[]
    for i in range(min(ds_responses_count, max_comparing)):
        one_score=compare_with_gpt(chatgpt_response,ds_responses.iloc[i])
        scores.append(one_score)
        print(one_score,' ', end='')
    scores=np.array(scores)
    min_score, max_score, mean_score=scores.min(), scores.max(),scores.mean()
    print(f' ]  min: {min_score},  max: {max_score},  mean: {mean_score} ')
    gpt_dataset.loc[current_ds_len]={
                'request': request,
                'label': label,
                'ds_responses_count': ds_responses_count,         
                'chatgpt_response': chatgpt_response,
                'scores': scores,
                'min_score': min_score,
                'max_score': max_score,
                'mean_score': mean_score
    }
    

In [31]:
def save_simulation_dataset(gpt_dataset=gpt_simulation):
    """Saves to csv the simulation dataset, using INDEX and Dataset lenght to make the filename

    Args:
        dataset (_type_, optional): _description_. Defaults to gpt_simulation.
    """
    ds_len = len(gpt_dataset)
    print()
    if ds_len:
        filename=GPT_SIMULATION_DATASET_FILENAME+str(ds_len).zfill(4)+'.csv'
        print(ds_len,' len dataset will be saved to ',filename)
        gpt_dataset.to_csv(filename)
        print('Dataset saved to ', filename)

### Ready to go!

## SIMULATION


Ok, we will start with entries, that have 3 or more original dataset responses
We will be doing 3 trials for each request

In [32]:
del gpt_simulation

gpt_simulation = pd.DataFrame(columns=['request', 'label', 'ds_responses_count',
                              'chatgpt_response', 'scores', 'min_score', 'max_score', 'mean_score'])

GPT_SIMULATION_DATASET_FILENAME = 'simulation_data/gpt_simulation_03_'

In [33]:
NUM_TRIALS=3

MAX_COMPARING=10

SAVE_ON_NUMBER=25

context_from_3_responses=context_unique[context_unique['count']>2]


In [34]:
sim_length=len(context_from_3_responses)

In [ ]:
## DONT RUN IT WE HAVE PARTIAL RESULTS OF IT

#    for request_number in range(sim_length):
# #     request=context_from_3_responses.iloc[request_number].iloc[0]
# #     label=get_label_from_chatGPT(request)
# #     print(f'context[{request_number}][{request[:150]}] (L:{label}) :')
                                          
# #     for trial in range(NUM_TRIALS):
# #         run_one_request(request,label,gpt_dataset=gpt_simulation,max_comparing=MAX_COMPARING)
    
# #     if (request_number % SAVE_ON_NUMBER==0)and(request_number>0):
# #         save_simulation_dataset(gpt_dataset=gpt_simulation)   
# #         print()
# #     print()
#          

context[0][A few years ago I was making love to my wife when for no known reason I lost my erection,    Now I'm In my early 30s and my problem has become more an] (L:Topic: Sexual Dysfunction
Mental Disorder: Depression) :
SIM (0) [A few years ago I wa]  ->  {I'm sorry to hear that you are going through this, and it's important to know that you are not alone in facing this issue. Erectile dysfunction can ha} :: [70  75  85  25  74  70   ]  min: 25,  max: 85,  mean: 66.5 
SIM (1) [A few years ago I wa]  ->  {I'm sorry to hear that you are going through this difficult situation. Erectile dysfunction (ED) can have various physical and psychological causes. I} :: [74  75  85  60  85  74   ]  min: 60,  max: 85,  mean: 75.5 
SIM (2) [A few years ago I wa]  ->  {I'm sorry to hear that you're going through this challenging situation. Erectile dysfunction can have both physical and psychological causes, and it's} :: [72  70  85  20  85  85   ]  min: 20,  max: 85,  mean: 69.5 

context[1][After 

BadRequestError: Error code: 400 - {'error': {'message': "This model's maximum context length is 16385 tokens. However, your messages resulted in 23181 tokens. Please reduce the length of the messages.", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}

In [36]:
# gpt_simulation

In [130]:
# save_simulation_dataset(gpt_dataset=gpt_simulation)


147  len dataset will be saved to  simulation_data/gpt_simulation_03_0147.csv
Dataset saved to  simulation_data/gpt_simulation_03_0147.csv


## We have got an error with message length in 23181 tokens. This model's maximum context length is 16385 tokens. 

Let's rewrite a comparing function and keep what we have got so far

First we need to count tokens

In [131]:
# !pip install --upgrade tiktoken

     ---------------------------------------- 0.0/42.0 kB ? eta -:--:--
     ------------------ ------------------- 20.5/42.0 kB 640.0 kB/s eta 0:00:01
     -------------------------------------  41.0/42.0 kB 991.0 kB/s eta 0:00:01
     -------------------------------------- 42.0/42.0 kB 503.7 kB/s eta 0:00:00
   ---------------------------------------- 0.0/798.7 kB ? eta -:--:--
   ---- ----------------------------------- 92.2/798.7 kB 2.6 MB/s eta 0:00:01
   ------------------- -------------------- 389.1/798.7 kB 4.8 MB/s eta 0:00:01
   ---------------------------------------- 798.7/798.7 kB 7.2 MB/s eta 0:00:00
   ---------------------------------------- 0.0/269.5 kB ? eta -:--:--
   --------------------------------------- 269.5/269.5 kB 16.2 MB/s eta 0:00:00
   ---------------------------------------- 0.0/62.6 kB ? eta -:--:--
   ---------------------------------------- 62.6/62.6 kB 3.3 MB/s eta 0:00:00
   ---------------------------------------- 0.0/99.9 kB ? eta -:--:--
   ------

In [44]:
import tiktoken

def num_tokens_from_string(string: str, encoding_name:str="cl100k_base") -> int:
    encoding = tiktoken.get_encoding(encoding_name)
    num_tokens = len(encoding.encode(string))
    return num_tokens

print(num_tokens_from_string("Hello world, let's test tiktoken.", "cl100k_base"))

# NOW FUNCTION IN THE BEGINNING OF NOTEBOOK IS CORRECT, BUT IS WAS LIKE THIS BEFORE, and this caused the error
# 

def generate_comparing_prompt(response1: str, response2: str, comparing_template=COMPARING_PROMPT_TEMPLATE_2) -> str:
    """Insert into COMPARING_PROMPT_TEMPLATE given request and responses

    Args:
        response1 (str): response1
        response2 (str): response2

    Returns:
        str: Prompt to send to ChatGRT for comparing
    """
    result=comparing_template.replace('<<<PHRASE1###>>>', response1).replace('<<<PHRASE2###>>>', response2)
    
    return result

9


In [50]:
r_num=49 
request=context_from_3_responses.iloc[r_num].iloc[0]
# label=get_label_from_chatGPT(request)
print(f'context[{r_num}][{request[:150]}] (L:) :')

context[49][I get so much anxiety, and I don’t know why. I feel like I can’t do anything by myself because I’m scared of the outcomes.] (L:) :


In [51]:
chatgpt_response=speak_with_gpt(request, temperature=SPEAK_TEMPERATURE)
ds_responses=conversation_dataset[conversation_dataset['context']==request]['response']

In [58]:
for i in range(len(ds_responses)):
    print(num_tokens_from_string(generate_comparing_prompt(chatgpt_response,ds_responses.iloc[i])))

523
469
362
449
23166


We will rewrite the function to make comparing prompts
We will truncate dataset_response if it exceeds limit 

In [63]:
def num_tokens_from_string(string: str, encoding_name:str="cl100k_base") -> int:
    encoding = tiktoken.get_encoding(encoding_name)
    num_tokens = len(encoding.encode(string))
    return num_tokens

# This is corrected one

def generate_comparing_prompt(response1: str, response2: str, comparing_template=COMPARING_PROMPT_TEMPLATE_2) -> str:
    """Insert into COMPARING_PROMPT_TEMPLATE given request and responses

    Args:
        response1 (str): response1
        response2 (str): response2

    Returns:
        str: Prompt to send to ChatGRT for comparing
    """
    result=comparing_template.replace('<<<PHRASE1###>>>', response1).replace('<<<PHRASE2###>>>', response2)
    while num_tokens_from_string(result)>14000:
        response2=response2[:-abs(num_tokens_from_string(result)-14000)]
        result=comparing_template.replace('<<<PHRASE1###>>>', response1).replace('<<<PHRASE2###>>>', response2)
    return result
    

In [68]:
for i in range(len(ds_responses)):
    print(num_tokens_from_string(generate_comparing_prompt(chatgpt_response,ds_responses.iloc[i])))

523
469
362
449
14000


In [69]:
for i in range(len(ds_responses)):
    print(compare_with_gpt(chatgpt_response,ds_responses.iloc[i]))

78
35
70
70
30


### Problem solved, restarting from request_number=49


In [70]:
gpt_simulation

Empty DataFrame
Columns: [request, label, ds_responses_count, chatgpt_response, scores, min_score, max_score, mean_score]
Index: []

In [71]:
gpt_simulation=pd.read_csv('simulation_data/gpt_simulation_03_0147.csv')

In [76]:
gpt_simulation

Unnamed: 0                                            request  \
0             0  A few years ago I was making love to my wife w...   
1             1  A few years ago I was making love to my wife w...   
2             2  A few years ago I was making love to my wife w...   
3             3  After 40 years of being straight, how could I ...   
4             4  After 40 years of being straight, how could I ...   
..          ...                                                ...   
142         142  I found out my boyfriend takes anti-depression...   
143         143  I found out my boyfriend takes anti-depression...   
144         144  I get angry and act out. It scares my wife and...   
145         145  I get angry and act out. It scares my wife and...   
146         146  I get angry and act out. It scares my wife and...   

                                                 label  ds_responses_count  \
0    Topic: Sexual Dysfunction\nMental Disorder: De...                   6   
1    Topic: Sexual Dysfunction\nMental Disorder: De...                   6   
2    Topic: Sexual Dysfunction\nMental Disorder: De...                   6   
3                           Sexuality, Identity Crisis                   6   
4                           Sexuality, Identity Crisis                   6   
..                                                 ...                 ...   
142                    Relationship Issues, Depression                   3   
143                    Relationship Issues, Depression                   3   
144                               Anger, Family Stress                   3   
145                               Anger, Family Stress                   3   
146                               Anger, Family Stress                   3   

                                      chatgpt_response               scores  \
0    I'm sorry to hear that you are going through t...  [70 75 85 25 74 70]   
1    I'm sorry to hear that you are going through t...  [74 75 85 60 85 74]   
2    I'm sorry to hear that you're going through th...  [72 70 85 20 85 85]   
3    Sexual orientation can be fluid and can change...  [85 70 90 90 70 65]   
4    Sexual orientation can be complex and fluid, a...  [85 40 89 90 75 63]   
..                                                 ...                  ...   
142  I'm sorry to hear that you're going through a ...           [80 70 70]   
143  It's understandable to feel concerned about yo...           [85 70 65]   
144  I'm sorry to hear that you're struggling with ...           [85 70 85]   
145  It's important to recognize when your anger is...           [85 70 75]   
146  It's important to recognize and address these ...           [75 74 70]   

     min_score  max_score  mean_score  
0           25         85   66.500000  
1           60         85   75.500000  
2           20         85   69.500000  
3           65         90   78.333333  
4           40         90   73.666667  
..         ...        ...         ...  
142         70         80   73.333333  
143         65         85   73.333333  
144         70         85   80.000000  
145         70         85   76.666667  
146         70         75   73.000000  

[147 rows x 9 columns]

In [78]:
GPT_SIMULATION_DATASET_FILENAME = 'simulation_data/gpt_simulation_04_'

def save_simulation_dataset(gpt_dataset=gpt_simulation):
    """Saves to csv the simulation dataset, using INDEX and Dataset lenght to make the filename

    Args:
        dataset (_type_, optional): _description_. Defaults to gpt_simulation.
    """
    ds_len = len(gpt_dataset)
    print()
    if ds_len:
        filename=GPT_SIMULATION_DATASET_FILENAME+str(ds_len).zfill(4)+'.csv'
        print(ds_len,' len dataset will be saved to ',filename)
        gpt_dataset.to_csv(filename)
        print('Dataset saved to ', filename)
        

Now below I am starting from 49, because we already done the work for first 48. If you want to run this simulation from the start, please run on a full range

In [79]:
NUM_TRIALS=3

MAX_COMPARING=10

SAVE_ON_NUMBER=25

context_from_3_responses=context_unique[context_unique['count']>2]


In [80]:
for request_number in range(49,sim_length):
    request=context_from_3_responses.iloc[request_number].iloc[0]
    label=get_label_from_chatGPT(request)
    print(f'context[{request_number}][{request[:120]}] (L:{label}) :')
                                          
    for trial in range(NUM_TRIALS):
        run_one_request(request,label,gpt_dataset=gpt_simulation,max_comparing=MAX_COMPARING)
    
    if (request_number % SAVE_ON_NUMBER==0)and(request_number>0):
        save_simulation_dataset(gpt_dataset=gpt_simulation)   
        print()
    print()

context[49][I get so much anxiety, and I don’t know why. I feel like I can’t do anything by myself because I’m scared of the outcome] (L:Anxiety, Fear) :
SIM (147) [I get so much a]  ->  {I'm sorry to hear that you're experiencing anxiety. It's important to remember that it's okay to fee} :: [70  70  85  70  20   ]  min: 20,  max: 85,  mean: 63.0 
SIM (148) [I get so much a]  ->  {I'm sorry to hear that you're experiencing anxiety. It's important to remember that anxiety is a com} :: [80  35  85  85  25   ]  min: 25,  max: 85,  mean: 62.0 
SIM (149) [I get so much a]  ->  {I'm sorry to hear that you're feeling this way. It's important to remember that anxiety is a common } :: [70  70  70  70  45   ]  min: 45,  max: 70,  mean: 65.0 

context[50][I have PTSD. The side effects are really bad and have impacts on sex, anger and my relationships. I even lost my job mon] (L:PTSD, Job Loss) :
SIM (150) [I have PTSD. Th]  ->  {I'm sorry to hear that you're going through such a challenging time.

In [81]:
save_simulation_dataset()


621  len dataset will be saved to  simulation_data/gpt_simulation_04_0621.csv
Dataset saved to  simulation_data/gpt_simulation_04_0621.csv


## Simulation dataset analysis

continued in the [simulation analysis notebook text](notebook_simulation_dataset_analysis.ipynb)